File: pt-4/156.binary-tree-upside-down.java
/**
 * Algorithm:
 * Iteratively rotate the tree along the left spine. The old parent becomes the
 * new right child, and the old parent's right child becomes the new left child.
 *
 * Complexity:
 * Time O(h) for the left spine length, space O(1).
 */
class Solution {
    public TreeNode upsideDownBinaryTree(TreeNode root) {
        TreeNode parent = null;
        TreeNode parentRight = null;
        TreeNode cur = root;
        while (cur != null) {
            TreeNode nextLeft = cur.left;
            cur.left = parentRight;
            parentRight = cur.right;
            cur.right = parent;
            parent = cur;
            cur = nextLeft;
        }
        return parent;
    }
}

File: pt-4/156.binary-tree-upside-down.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

// Definition for a binary tree node.
// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };


class Solution {
public:
    TreeNode* upsideDownBinaryTree(TreeNode* root) {
        /*
        Approach: iteratively rotate the left spine. The old parent becomes the
        new right child, and the old right sibling becomes the new left child.
        Track parent and parentRight while moving down original left pointers.

        Complexity: O(h) time where h is left-spine length, O(1) space.
        */
        TreeNode* parent = nullptr;
        TreeNode* parentRight = nullptr;
        TreeNode* cur = root;
        while (cur) {
            TreeNode* nextLeft = cur->left;
            cur->left = parentRight;
            parentRight = cur->right;
            cur->right = parent;
            parent = cur;
            cur = nextLeft;
        }
        return parent;
    }
};
